tests/snap_test.go

```
@@ -215,22 +227,35 @@ func TestWifiMatterCommander(t *testing.T) {
 	var stdout, stderr string
 	var err error
 
+	start := time.Now()
+
+	t.Cleanup(func() {
+		utils.SnapStop(t, snapMatterPiGPIO)
+
+		if env.Teardown() {
+			// Remove snaps, ignore errors during removal
+			utils.SnapRemove(nil, chipToolSnap)
+			// snapMatterPiGPIO is not removed here, as that is handled by the teardown function
+		}
```

Each test should schedule to revert anything it has done, as it happens, not in advance. This is because it is not yet clear if for example Chip Tool will succeed in installation (needing a removal), or commander will succeed to start (needing a stop).

The logic should be similar to deferred statements, except that they get executed on test failure rather than return to the caller. The functions get added to a stack and executed in LIFO order, reverting the set-up in the reverse order.

---

I agree, but this is the logic that was originally used in these tests. I assume it is because gpio-commander is modified for mocked gpio, and we only want to do that, and the installation once.

---

I re-read your comment. I think I understand now. I'll break this one cleanup into individual ones.